c/tskit/trees.c

```diff
@@ -2124,6 +2124,580 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    tsk_memset(&AB_samples, 0, sizeof(AB_samples));
+
+    if (weights == NULL || norm == NULL || result_tmp == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    ret = tsk_bit_array_init(&ss_A_samples, num_samples, 1);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_bit_array_init(&ss_B_samples, num_samples, 1);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_bit_array_init(&ss_AB_samples, num_samples, 1);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_bit_array_init(&AB_samples, num_samples, 1);
+    if (ret != 0) {
+        goto out;
+    }
+
+    for (mut_a = polarised_val; mut_a < num_a_alleles; mut_a++) {
+        result_tmp_row = GET_2D_ROW(result_tmp, row_len, mut_a);
+        for (mut_b = polarised_val; mut_b < num_b_alleles; mut_b++) {
+            tsk_bit_array_get_row(site_a_state, mut_a, &A_samples);
+            tsk_bit_array_get_row(site_b_state, mut_b, &B_samples);
+            tsk_bit_array_intersect(&A_samples, &B_samples, &AB_samples);
+            for (k = 0; k < state_dim; k++) {
+                tsk_bit_array_get_row(sample_sets, k, &ss_row);
+                hap_weight_row = GET_2D_ROW(weights, 3, k);
+
+                tsk_bit_array_intersect(&A_samples, &ss_row, &ss_A_samples);
+                tsk_bit_array_intersect(&B_samples, &ss_row, &ss_B_samples);
+                tsk_bit_array_intersect(&AB_samples, &ss_row, &ss_AB_samples);
+
+                w_AB = tsk_bit_array_count(&ss_AB_samples);
+                w_A = tsk_bit_array_count(&ss_A_samples);
+                w_B = tsk_bit_array_count(&ss_B_samples);
+
+                hap_weight_row[0] = (double) w_AB;
+                hap_weight_row[1] = (double) (w_A - w_AB); // w_Ab
+                hap_weight_row[2] = (double) (w_B - w_AB); // w_aB
+            }
+            ret = f(state_dim, weights, result_dim, result_tmp_row, f_params);
+            if (ret != 0) {
+                goto out;
+            }
+            ret = norm_f(state_dim, weights, num_a_alleles - polarised_val,
+                num_b_alleles - polarised_val, norm, f_params);
+            if (ret != 0) {
+                goto out;
+            }
+            for (k = 0; k < state_dim; k++) {
+                result[k] += result_tmp_row[k] * norm[k];
+            }
+            result_tmp_row += state_dim; // Advance to the next column
+        }
+    }
+
+out:
+    tsk_safe_free(weights);
+    tsk_safe_free(norm);
+    tsk_safe_free(result_tmp);
+    tsk_bit_array_free(&ss_A_samples);
+    tsk_bit_array_free(&ss_B_samples);
+    tsk_bit_array_free(&ss_AB_samples);
+    tsk_bit_array_free(&AB_samples);
+    return ret;
+}
+
+static int
+get_all_mutation_samples(const tsk_treeseq_t *self, const tsk_size_t *site_offsets,
+    tsk_size_t *num_alleles, tsk_bit_array_t *allele_samples)
+{
+    int ret = 0;
+    tsk_tree_t tree;
+
+    ret = tsk_tree_init(&tree, self, TSK_TS_INIT_BUILD_INDEXES | TSK_NO_SAMPLE_COUNTS);
```

Comment:
``TSK_TS_INIT_BUILD_INDEXES``  isn't a valid option for tree_init.

Reply:
To me the ``get_mutation_samples`` function is more complicated than it needs to be, and it would be clearer to merge it into this function (as it does very little at the moment)

Reply:
I did a refactor of all of this as you've suggested. It seems much simpler. I think with some of the suggestions about the bit array refactor, this code will be simplified even more.